docs/developer/highlevel_design.rst

```diff
@@ -0,0 +1,220 @@
+* **Preprocess** - this stage allows the user to specify the filter/functions to be applied to the data and the order in which to apply them prior to the actual reconstruction
+* **Reconstruction** - This stage performs the actual reconstruction based on the intermediate results established during the preprocessing stage.
+* **Visualization** - This stage allows the user to interactively view and explore the result of the reconstruction
+
+The downloading stage is always presented to the user to either fill out the required data or ensure the existing data is correct.
+However, when the user chooses to run a non-interactive reconstruction the software performs the reconstruction process based on the configuration data and a predefined workflow that has been specified in the json configuration file.
+
+Capabilities
+------------
+
+The iMars3D UI Application provides the user the following capabilities:
+
+* Interactively specify the configuration information required for performing a reconstruction (configuration data)
+* Interactively perform a reconstruction
+* Interactively specify the workflow to execute to perform a reconstruction
+* Launch a non-interactive reconstruction process and monitor its progress
+
+Structure
+---------
+
+The iMars3D UI application is comprised of:
+
+* The Front End (UI)
+
+    * Stages - Stages group and control the user interaction
+    * Widgets - There is a widget for each filter/function to handle the user interaction with the specific filter/function.
+
+* The Backend
+
+    * iMars3D Filter/Function Library
+    * iMars3D Interactive Workflow Engine
+
+
+iMars3D Auto Reconstruction Application
+=======================================
+
+iMars3D auto reconstruction application and the application to be used when performing auto reconstruction.
+
+.. note:: Auto reconstruction is equivalent to auto reduction. It is invoked in the same way automatically after data acquisition.
+
+Auto reconstruction takes an already specified configuration json file which contains all the information required to run a reconstruction process.
+
+Structure
+---------
+
+The iMars3D auto reconstruction application si comprised of:
+
+* The Backend
+
+    * iMars3D Filter/Function Library
+    * IMars3D non-interactive Workflow Engine
+
+This application could also be run from the command line.
+
+For auto reconstruction a wrapper script is required.
+The system that invokes the auto reconstruction assumes an interface that takes the location of the raw ct_scans.
+
+iMars3D Library
+===============
+
+The iMars3D (filter/function) library is simply the set of filters and functions involved in performing a reconstruction process.
+The library is expected to be used within a python application.
+It is the application that keeps track of the state of the reconstruction and maintains the intermediate results as required to transition from one filter/function to the next.
```

Comment:
I look at this "auto reconstruction" capability as another iMars3D application.  If this were compiled code we would have an executable that is run.  The engine does not exist on its own, but needs to be bundled with the library of filters/function it can run as it attempts to execute a workflow.  So perhaps it is semantics but the engine is just an engine and incomplete on its own.

Reply:
It could be any "application" that uses the library filter/functions.  We have the auto reduction app (engine) and the UI app (also with an engine).  A user could create their own app and is the library.  IMO, the important  thing is that the library itself does not maintain state